crates/rome_js_analyze/src/analyzers/complexity/no_confusing_void_type.rs

```diff
@@ -0,0 +1,281 @@
+    /// let bar = 1 as unknown as void;
+    /// let baz = 1 as unknown as void | string;
+    /// ```
+    ///
+    /// ### Valid
+    ///
+    /// ```ts
+    /// function foo(): void {};
+    /// ```
+    ///
+    /// ## Options
+    ///
+    /// ### allowInGenericTypeArguments
+    ///
+    /// This option lets you control if void can be used as a valid value for generic type parameters.
+    ///
+    /// This option is `true` by default.
+    ///
+    /// The following patterns are considered warnings with `{ allowInGenericTypeArguments: false }`:
+    ///
+    /// ```ts
+    /// logAndReturn<void>(undefined);
+    ///
+    /// let voidPromise: Promise<void> = new Promise<void>(() => {});
+    /// let voidMap: Map<string, void> = new Map<string, void>();
+    /// ```
+    ///
+    /// ### allowAsThisParameter
+    ///
+    ///
+    ///
+    ///
+    /// This option allows specifying a this parameter of a function to be void when set to true. This pattern can be useful to explicitly label function types that do not use a this argument. See the TypeScript docs for more information.
+    /// This option is `false` by default.
+    ///
+    /// The following patterns are considered warnings with `{ allowAsThisParameter: false }` but valid with `{ allowAsThisParameter: true }`:
+    ///
+    /// ```ts
+    /// function doThing(this: void) {}
+    /// class Example {
+    ///     static helper(this: void) {}
+    ///     callback(this: void) {}
+    /// }
+    /// ```
+    ///
+    pub(crate) NoConfusingVoidType {
+        version: "1.0.0",
+        name: "noConfusingVoidType",
+        recommended: false,
+    }
+}
+
+/// rule options for `noConfusingVoidType`
+#[derive(Debug, Clone, Deserialize, Serialize, Bpaf)]
+#[cfg_attr(feature = "schema", derive(JsonSchema))]
+#[serde(rename_all = "camelCase", deny_unknown_fields)]
+pub struct NoConfusingVoidTypeOptions {
+    /// This option lets you control if void can be used as a valid value for generic type parameters.
+    allow_in_generic_type_arguments: bool,
+    /// This option allows specifying a this parameter of a function to be void when set to true. This pattern can be useful to explicitly label function types that do not use a this argument.
+    allow_as_this_parameter: bool,
+}
+
+impl NoConfusingVoidTypeOptions {
+    const KNOWN_KEYS: &'static [&'static str] =
+        &["allowInGenericTypeArguments", "allowAsThisParameter"];
+}
+
+impl Default for NoConfusingVoidTypeOptions {
+    fn default() -> Self {
+        Self {
+            allow_in_generic_type_arguments: true,
+            allow_as_this_parameter: false,
+        }
+    }
+}
+
+type Language = <AnyTsType as AstNode>::Language;
+
+// We only focus on union type
+pub enum VoidTypeIn {
+    Union,
+    Unknown,
+}
+
+impl Rule for NoConfusingVoidType {
+    type Query = Ast<AnyTsType>;
+
+    type State = VoidTypeIn;
+
+    type Signals = Option<Self::State>;
+
+    type Options = NoConfusingVoidTypeOptions;
```

Comment:
```suggestion
    type Query = Ast<AnyTsType>;
    type State = VoidTypeIn;
    type Signals = Option<Self::State>;
    type Options = NoConfusingVoidTypeOptions;
```

To be more consistent with other implementations, I suggest removing blank lines here.